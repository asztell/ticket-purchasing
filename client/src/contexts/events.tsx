import { createContext, useEffect, useMemo, useState } from "react";

export type Event = {
  id: string;
  name: string;
  location: string;
  ISODate: string;
  price: number;
};

export type EventsContextType = {
  events: Event[];
  error: string | null;
};

export const EventsContext = createContext<EventsContextType>(
  {} as EventsContextType
);

export function EventsProvider({ children }: { children: React.ReactNode }) {
  const [events, setEvents] = useState<Event[]>([]);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    // const url = `${document.location.origin}/events`;
    const url = `${process.env.REACT_APP_API}/events`;
    console.log("url", url);
    fetch(url)
      .then((res) => {
        if (!res.ok) {
          throw res;
        }
        return res.json();
      })
      .then((result) => {
        console.log("events", result);
        setEvents(result);
      })
      .catch((error) => {
        setError(error);
      });
  }, [setEvents]);

  const value = useMemo(() => {
    return {
      events,
      error,
    };
  }, [events, error]);

  return (
    <EventsContext.Provider value={value}>{children}</EventsContext.Provider>
  );
}
