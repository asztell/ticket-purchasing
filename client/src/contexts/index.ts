export * from './events'
export * from './ticketPurchasing'
