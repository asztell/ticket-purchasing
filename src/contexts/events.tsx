import { createContext, useEffect, useMemo, useState } from "react";

export type Event = {
  id: string;
  name: string;
  location: string;
  ISODate: string;
  price: number;
};

export type EventsContextType = {
  events: Event[];
  error: string | null;
};

export const EventsContext = createContext<EventsContextType>(
  {} as EventsContextType
);

export function EventsProvider({ children }: { children: React.ReactNode }) {
  const [events, setEvents] = useState<Event[]>([]);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    fetch(`http://localhost:8080/events`)
      .then((res) => {
        if (!res.ok) {
          throw res;
        }
        return res.json();
      })
      .then((result) => {
        setEvents(result);
      })
      .catch((error) => {
        setError(error);
      });
  }, [setEvents]);

  const value = useMemo(() => {
    return {
      events,
      error,
    };
  }, [events, error]);

  return (
    <EventsContext.Provider value={value}>{children}</EventsContext.Provider>
  );
}
