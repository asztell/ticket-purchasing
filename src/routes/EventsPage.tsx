import { useContext } from "react";
import classnames from "classnames";
import { useAuth0 } from "@auth0/auth0-react";
import {
  Events,
  Quantity,
  LinkButton,
  NavigationContainer,
} from "../components";
import { useQueryParams, getQueryParams } from "../hooks/useQueryParams";
// import { TicketPurchasingContext } from "../contexts";

export function EventsPage() {
  // const { selectedEvent, ticketsCounter } = useContext(TicketPurchasingContext);
  const [queryParams] = useQueryParams();
  const { eventId, ticketsCount } = getQueryParams(queryParams);
  const disabledLinkButton = !eventId || !ticketsCount;
  const { isAuthenticated } = useAuth0();
  const toCkeckoutPage = isAuthenticated ? "/checkout" : "/login";

  const eventsPageToHomePageClassName = classnames(
    "Events-Page-To-Home-Page",
    "Left-Link-Button"
  );
  const eventsPageToCheckoutPageClassName = classnames(
    "Events-Page-To-Checkout-Page",
    "Right-Link-Button"
  );

  return (
    <div className="Events-Page">
      <div>
        <Events />
        <Quantity />
      </div>
      {/* <NavigationContainer>
        <LinkButton
          to="/"
          label="< Home Page"
          className={eventsPageToHomePageClassName}
        />
        <LinkButton
          // to="/checkout"
          to={toCkeckoutPage}
          disabled={disabledLinkButton}
          label="Next >"
          className={eventsPageToCheckoutPageClassName}
        />
      </NavigationContainer> */}
    </div>
  );
}
