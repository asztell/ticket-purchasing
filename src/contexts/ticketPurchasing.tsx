import { createContext, useCallback, useMemo, useState } from "react";
import { Event } from "./events";

export type CardInfoType = {
  nameOnCard: string;
  cardNumber: string;
  cardType: string;
  securityCode: string;
  securityCodeValid: boolean;
  expirationDate: string;
  expirationDateValid: boolean;
};

export type TicketPurchasingContextType = {
  selectedEvent: Event | undefined;
  ticketsCounter: number;
  cardInfo: CardInfoType;
  updateSelectedEvent: (selectedEvent: Event) => void;
  updateTicketsCounter: (ticketsCounter: number) => void;
  updateCardInfo: (cardInfo: CardInfoType) => void;
};

export const TicketPurchasingContext =
  createContext<TicketPurchasingContextType>({} as TicketPurchasingContextType);

export function TicketPurchasingProvider({
  children,
}: {
  children: React.ReactNode;
}) {
  const [ticketsCounter, setTicketsCounter] = useState(0);
  const [selectedEvent, setSelectedEvent] = useState<Event | undefined>(
    undefined
  );
  const [cardInfo, setCardInfo] = useState({
    nameOnCard: "",
    cardNumber: "",
    cardType: "",
    securityCode: "",
    securityCodeValid: false,
    expirationDate: "",
    expirationDateValid: false,
  });

  const updateTicketsCounter = useCallback((ticketsCounter: number) => {
    setTicketsCounter(
      (prevTicketsCounter) => prevTicketsCounter + ticketsCounter
    );
  }, []);

  const updateSelectedEvent = useCallback((selectedEvent: Event) => {
    setSelectedEvent(selectedEvent);
  }, []);

  const updateCardInfo = useCallback((cardInfo: CardInfoType) => {
    const { expirationDate, cardNumber, securityCode } = cardInfo;
    const cardType = validateCardNumber(cardNumber);
    const securityCodeValid =
      securityCode.match(new RegExp("^[0-9]{3,4}$")) !== null;
    const expirationDateValid = validateExpirationDate(expirationDate);
    setCardInfo({
      ...cardInfo,
      cardType,
      securityCodeValid,
      expirationDateValid,
    });
  }, []);

  const value = useMemo(() => {
    return {
      selectedEvent,
      ticketsCounter,
      cardInfo,
      updateSelectedEvent,
      updateTicketsCounter,
      updateCardInfo,
    };
  }, [
    selectedEvent,
    ticketsCounter,
    cardInfo,
    updateSelectedEvent,
    updateTicketsCounter,
    updateCardInfo,
  ]);
  return (
    <TicketPurchasingContext.Provider value={value}>
      {children}
    </TicketPurchasingContext.Provider>
  );
}

export function validateCardNumber(cardNumber: string) {
  const visa = new RegExp("^4[0-9]{12}(?:[0-9]{3})?$");
  if (cardNumber.match(visa) !== null) return "Visa";
  const mastercard = new RegExp(
    "^5[1-5][0-9]{14}|^(222[1-9]|22[3-9]\\d|2[3-6]\\d{2}|27[0-1]\\d|2720)[0-9]{12}$"
  );
  if (cardNumber.match(mastercard) !== null) return "Mastercard";
  const amex = new RegExp("^3[47][0-9]{13}$");
  if (cardNumber.match(amex)) return "American Express";
  const discover = new RegExp("^6(?:011|5[0-9]{2})[0-9]{12}$");
  if (cardNumber.match(discover)) return "Discover";
  const dinersClub = new RegExp("^3(?:0[0-5]|[68][0-9])[0-9]{11}$");
  if (cardNumber.match(dinersClub)) return "Diners Club";
  if (cardNumber.length === 0) return "";
  return "Invalid";
}

export function validateExpirationDate(expirationDate: string) {
  const expirationDateValid =
    expirationDate.match(
      new RegExp("^(0[1-9]|1[0-2])\\/?([0-9]{4}|[0-9]{2})$")
    ) !== null && new Date(expirationDate) > new Date();
  return expirationDateValid;
}
