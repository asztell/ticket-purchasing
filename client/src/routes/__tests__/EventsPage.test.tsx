import { render } from "@testing-library/react";
import { MemoryRouter as Router } from "react-router-dom";
import { EventsPage } from "../EventsPage";
import { TicketPurchasingContext, EventsContext } from "../../contexts";

describe("<EventsPage />", () => {
  test("render EventsPage", () => {
    const { container } = render(
      <Router>
        <EventsContext.Provider
          value={{
            events: [
              {
                name: "Test Event",
                location: "Test Location",
                ISODate: "2021-01-01",
                price: 100,
              },
            ],
            error: "",
          }}
        >
          <TicketPurchasingContext.Provider
            value={{
              selectedEvent: {
                name: "Test Event",
                location: "Test Location",
                ISODate: "2021-01-01",
                price: 100,
              },
              ticketsCounter: 1,
              cardInfo: {
                cardNumber: "1234123412341234",
                nameOnCard: "Test User",
                cardType: "Test Card",
                securityCode: "123",
                securityCodeValid: true,
                expirationDate: "01/01/2021",
                expirationDateValid: true,
              },
              termsOfUseChecked: true,
              updateSelectedEvent: () => {},
              updateTicketsCounter: () => {},
              updateCardInfo: () => {},
              updateTermsOfUseChecked: () => {},
            }}
          >
            <EventsPage />
          </TicketPurchasingContext.Provider>
        </EventsContext.Provider>
      </Router>
    );
    expect(container).toMatchSnapshot();
  });
});
