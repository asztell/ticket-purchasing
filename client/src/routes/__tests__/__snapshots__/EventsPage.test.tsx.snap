// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventsPage /> render EventsPage 1`] = `
<div>
  <div
    class="Events-Page"
  >
    <div>
      <div
        class="Events"
      >
        <h2>
          Events
        </h2>
        <div>
          <label
            for="event-select"
          >
            Choose an event:
          </label>
          <div
            class="Event-Card Selected-Event-Card Event"
          >
            <div
              class="Event-Name"
            >
              Test Event
            </div>
            <div
              class="Event-Date"
            >
              Thursday • 12/31/2020 • 7:00:00 PM
            </div>
            <div
              class="Event-Location"
            >
              Test Location
            </div>
            <div
              class="Event-Price"
            >
              Price: $100.00
            </div>
          </div>
        </div>
      </div>
      <div
        class="Quantity"
      >
        <h2>
          Quantity
        </h2>
        <div
          class="Container"
        >
          <button>
            -
          </button>
          <span
            class="Value"
          >
            1
          </span>
          <button>
            +
          </button>
        </div>
      </div>
    </div>
    <div
      class="Navigation-Container"
    >
      <div>
        <button
          class="Link-Button Events-Page-To-Home-Page Left-Link-Button"
        >
          &lt; Home Page
        </button>
      </div>
      <div>
        <button
          class="Link-Button Events-Page-To-Checkout-Page Right-Link-Button"
        >
          Next &gt;
        </button>
      </div>
    </div>
  </div>
</div>
`;
